homeassistant/components/ecobee/switch.py

```diff
@@ -0,0 +1,105 @@
+"""Support for using switch with ecobee thermostats."""
+
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from dataclasses import dataclass
+from datetime import datetime
+import logging
+from typing import Any
+
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import EcobeeData
+from .const import DOMAIN
+from .entity import EcobeeBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+DATE_FORMAT = "%Y-%m-%d %H:%M:%S"
+
+
+@dataclass(frozen=True, kw_only=True)
+class EcobeeSwitchEntityDescription(SwitchEntity):
+    """Class describing Ecobee switch entities."""
+
+    ecobee_setting_key: str
+    set_fn: Callable[[EcobeeData, int, int], Awaitable]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the ecobee thermostat switch entity."""
+    data: EcobeeData = hass.data[DOMAIN]
+    _LOGGER.debug("Adding ventilators 20 min switch if present")
+
+    async_add_entities(
+        (
+            EcobeeVentilator20MinSwitch(data, index)
+            for index, thermostat in enumerate(data.ecobee.thermostats)
+            if thermostat["settings"]["ventilatorType"] != "none"
+        ),
+        True,
+    )
+
+
+class EcobeeVentilator20MinSwitch(EcobeeBaseEntity, SwitchEntity):
+    """A Switch class, representing 20 min timer for an ecobee thermostat with ventilator attached."""
+
+    def __init__(
+        self,
+        data: EcobeeData,
+        thermostat_index: int,
+    ) -> None:
+        """Initialize ecobee ventilator platform."""
+        super().__init__(data, thermostat_index)
+        self._attr_unique_id = "ventilator_20m_timer"
+        self._attr_native_value = False
+        self.update_without_throttle = False
+
+    @property
+    def is_on(self) -> bool:
+        """Get the latest state from the thermostat."""
+        return self._attr_native_value
+
+    async def async_update(self) -> None:
+        """Get the latest state from the thermostat."""
+
+        if self.update_without_throttle:
+            await self.data.update(no_throttle=True)
+            self.update_without_throttle = False
+        else:
+            await self.data.update()
+
+        ventilatorOffDateTime = self.thermostat["settings"]["ventilatorOffDateTime"]
+
+        time_zone_delay = datetime.strptime(
+            self.thermostat["utcTime"], DATE_FORMAT
+        ) - datetime.strptime(self.thermostat["thermostatTime"], DATE_FORMAT)
+
+        self._attr_native_value = (
+            ventilatorOffDateTime is not None
+            and ventilatorOffDateTime != ""
+            and datetime.strptime(ventilatorOffDateTime, DATE_FORMAT) + time_zone_delay
+            >= datetime.now()
+        )
+
```

Comment:
a better location for this, one. Ore question.  

Can this be simplified to compare ventilatorOffDateTime against thermostatTime?  And avoid the whole time zone delay calculation?

Reply:
Yeah, I thought about this, I gave it a try. 

I works but, because the Ecobee data is updated every 3 min it may take up to 3 min to get the accurate status

Exemple 
refresh 1:
thermostatTime: 12:12:30
ventilatorOffDateTime: 12:12:45

180 sec later (for that entire time the thermostatTime is not updating)

thermostatTime: 12:15:30
ventilatorOffDateTime: 12:12:45

So the switch stay on for 2:45 too long. 


---

We can go with this approach, code is actually simpler. But the data is a bit more stale 

Your call, I don't have a strong opinion on this one

Reply:
I see, while the update function gets called periodically and it can evaluate the expression, not every update cycle gets the device data.

let's not change it.